rtt/os/macosx/fosi.h

```diff
@@ -237,24 +249,319 @@ extern "C"
     /* 		return -1; */
     /*     } */
 
-    // Mutex functions - support only those needed by TLSF
-	// opaque type
-    typedef struct rt_mutex_impl_t rt_mutex_impl_t;
-	// type created by tlsf.c (must work in C, not C++ code)
-    typedef rt_mutex_impl_t* rt_mutex_t;
-    int rtos_mutex_init(rt_mutex_t* m);
-    int rtos_mutex_destroy(rt_mutex_t* m);
-	int rtos_mutex_lock( rt_mutex_t* m);
-	int rtos_mutex_unlock( rt_mutex_t* m);
+    // Mutex functions
+
+
+    // Note: Apple's implementation of pthread_mutex_t does not have pthread_mutex_timedlock.
+    // Workaround: Pair the mutex with a condition variable that is signaled whenever the mutex is unlocked.
+    // This is also how boost::timed_mutex is implemented for the case BOOST_THREAD_USES_PTHREAD_TIMEDLOCK is
+    // not defined (cf. https://github.com/boostorg/thread/blob/boost-1.62.0/include/boost/thread/pthread/mutex.hpp#L181).
+    typedef struct {
+        pthread_mutex_t m;
+        pthread_cond_t cond;
+        bool is_locked;
+    } rt_mutex_t;
+    typedef struct {
+        pthread_mutex_t m;
+        pthread_cond_t cond;
+        bool is_locked;
+        pthread_t owner;
+        unsigned count;
+    } rt_rec_mutex_t;
+
+    #define _RTOS_INTERNAL_MUTEX_LOCK \
+      RTT_VERIFY(!(ret = pthread_mutex_lock((&(m->m))))); \
+      if (ret != 0) return ret
+
+    #define _RTOS_INTERNAL_MUTEX_UNLOCK_AND_RETURN \
+      RTT_VERIFY(!pthread_mutex_unlock(&(m->m))); \
+      return
+
+    static inline int rtos_mutex_init(rt_mutex_t* m)
+    {
+        pthread_mutexattr_t attr;
+        int ret = 0;
+
+        RTT_VERIFY(!(pthread_mutexattr_init(&attr)));
+        if (ret != 0) return ret;
+
+        // enable priority inheritance
+        // The mutex only protects the internal flags, so no use to enable priority inheritance.
+//        RTT_VERIFY(!(ret = pthread_mutexattr_setprotocol(&ma_t, PTHREAD_PRIO_INHERIT)));
+//        if (ret != 0) {
+//            RTT_VERIFY(!pthread_mutexattr_destroy(&attr));
+//            return ret;
+//        }
+
+        // initialize mutex
+        RTT_VERIFY(!(ret = pthread_mutex_init(&(m->m), &attr)));
+        if (ret != 0) {
+            RTT_VERIFY(!pthread_mutexattr_destroy(&attr));
+            return ret;
+        }
+        pthread_mutexattr_destroy(&attr);
+
+        // initialize condition variable (for rtos_mutex_lock_until)
+        RTT_VERIFY(!(ret = pthread_cond_init(&(m->cond), NULL)));
+        if (ret != 0) {
+            RTT_VERIFY(!pthread_mutex_destroy(&(m->m)));
+            return ret;
+        }
+
+        m->is_locked = false;
+        return ret;
+    }
+
+    static inline int rtos_mutex_destroy(rt_mutex_t* m )
+    {
+        RTT_VERIFY(!pthread_cond_destroy(&(m->cond)));
+        int ret = 0;
+        RTT_VERIFY(!(ret = pthread_mutex_destroy(&(m->m))));
+        return ret;
+    }
+
+    static inline int rtos_mutex_rec_init(rt_rec_mutex_t* m)
+    {
+        m->is_locked = false;
```

Comment:
Is there a reason for which you need to initialize `is_locked` here (which will be initialized anyway in `rtos_mutex_init`), and not `owner`? :thinking:

Reply:
No, copy and paste. The first implementation had a lot of copied code and did not call into `rtos_mutex_init()` from `rtos_mutex_rec_init()`. I will remove this assignment.